judge/models/contest.py

```diff
@@ -182,17 +186,32 @@ def can_see_scoreboard(self, user):
             return False
         if self.hide_scoreboard and not self.is_in_contest(user) and self.end_time > self._now:
             return False
+        if self.partially_hide_scoreboard and not self.is_finished_contest(user):
+            return False
         return True
 
     def can_see_full_scoreboard(self, user):
         if self.is_editable_by(user):
             return True
         if not self.is_accessible_by(user):
+            # Not accessible by user
             return False
         if not self.show_scoreboard:
             return False
+        if self.partially_hide_scoreboard and not self.is_finished_contest(user):
+            return False
         return True
 
+    def is_finished_contest(self, user):
+        if user.is_authenticated:
+            profile = user.profile
+            if profile:
+                for participation in ContestParticipation.objects.filter(virtual=0, user=profile) \
```

Comment:
This is an O(N) operation that can be done in O(1) with an addition to the filter.
Suggested replacement:
```py
participation = self.users.filter(virtual=ContestParticipation.LIVE, user=profile).first()
```
Note that `participation` is `None` if the participation does not exist.